### PR TITLE
fix(controller): reject query string or fragment in upstreamDefinitions URLs

### LINKED ISSUE #2068: upstreamDefinitions URLs do not support query strings or fragments
### Description

`upstreamDefinitions[].upstreams[].url` is intended to be host[:port] only — the base path is configured exclusively via `upstreamDefinitions[].basePath` (see #2064 and https://github.com/wso2/api-platform/discussions/369#discussioncomment-16021268).

A query string (`?foo=bar`) or fragment (`#section`) in such a URL is not supported: it has no effect. The upstream cluster is built from host:port only, so the query string and fragment are dropped and never reach the upstream. Today they are also not flagged — the deployment is accepted silently — so the user gets no signal that the value was ignored. (By contrast, a path in the URL is rejected with a 400.)

**Current behavior (verified)**

Verified against a running gateway via the integration test `gateway/it/features/upstream-url-format.feature`:

| `upstreamDefinitions` URL | Deploy | Result |
|---|---|---|
| `http://backend:9080?foo=bar` (+ `basePath: /alternate`) | accepted | query string dropped; routes to `/alternate/<op>` |
| `http://backend:9080#section` (+ `basePath: /alternate`) | accepted | fragment dropped; routes to `/alternate/<op>` |
| `http://backend:9080/foo?x=1` | 400 | rejected: "URL must not include a path" |

```yaml
upstreamDefinitions:
  - name: q-upstream
    basePath: /alternate
    upstreams:
      - url: http://backend:9080?foo=bar   # or http://backend:9080#section — silently ignored
```

**Possible resolutions**

- Reject a query string or fragment at deploy time with a 400, consistent with how a path is handled (clearest, unambiguous contract).
- Alternatively, accept but explicitly document that query strings and fragments in `upstreamDefinitions` URLs are ignored.

### Version

_No response_

### Related Issue

#2064

### PR DESCRIPTION
**Pull request by mehara-rothila**
## Purpose

`upstreamDefinitions[].upstreams[].url` is intended to be host[:port] only — an upstream definition's base path is configured exclusively via `upstreamDefinitions[].basePath` (per #2064 and https://github.com/wso2/api-platform/discussions/369#discussioncomment-16021268).
A path in such a URL is already rejected with a 400 (#2065), but a query string (`?foo=bar`) or fragment (`#section`) is silently accepted: the upstream cluster is built from host:port only, so the query string and fragment are dropped and never reach the upstream — with no signal to the user that the value was ignored.
Fixes #2068.
Related #2064.

## Goals

Reject a query string or fragment in an `upstreamDefinitions` URL with a 400 at deploy time, consistent with how an embedded path is already handled — giving the user a clear, unambiguous contract that these URLs are host[:port] only.

## Approach

- `pkg/config/api_validator.go`: in `validateUpstreamDefinitions`, immediately after the existing #2065 path check, reject a URL whose `RawQuery` is non-empty or that carries a bare `?` (`ForceQuery`) with "URL must not include a query string", and a URL with a non-empty `Fragment` with "URL must not include a fragment".
- The validator is the single deploy-time gate for `upstreamDefinitions` URLs; the downstream consumers (`pkg/transform/restapi.go`, `pkg/xds/translator.go`) run after validation and read only host/scheme, so no change is needed there.
- Direct `upstream.main`/`upstream.sandbox` URLs are unchanged — they legitimately carry a base path in the URL path and are validated separately (`validateUpstreamUrl`); query/fragment handling for those is outside the scope of this issue.

## User stories

As an API developer, when I mistakenly put a query string or fragment in an `upstreamDefinitions` URL, I get a clear 400 at deploy time instead of having the value silently dropped at runtime.

## Documentation

N/A — no new configuration is introduced; this tightens validation of an existing field to match the documented host[:port]-only contract.

## Automation tests

 - Unit tests
   > Added `TestValidateUpstreamDefinitions_URLWithQueryRejected` and `TestValidateUpstreamDefinitions_URLWithFragmentRejected`, mirroring the existing `TestValidateUpstreamDefinitions_URLWithPathRejected`. `go test ./pkg/config/` passes with no regression.
 - Integration tests
   > N/A — this is deploy-time input validation with no runtime behavior change.

## Security checks

 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? N/A (Go code)
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples

N/A

## Related PRs

#2065 (take upstreamDefinitions base path from basePath) introduced the path rejection this builds on; #2059 (dynamic-endpoint sandbox routing) is adjacent work on the same area.

## Test environment

Go (multi-module workspace); `go test ./pkg/config/`.

### CODE REVIEW

**Comment by mehara-rothila** [comment]
@coderabbitai review

File: gateway/gateway-controller/pkg/config/validator_test.go [hunk 1]
```diff
@@ -620,6 +620,72 @@ func TestValidateUpstreamDefinitions_URLWithPathRejected(t *testing.T) {
 	assert.Contains(t, errors[0].Message, "basePath")
 }
 
+// upstreamDefinitions URLs must be host[:port] only; a query string is dropped, not honored.
```
[the remark this answers is not in the document]
Comment (mehara-rothila) [reply gateway/gateway-controller/pkg/config/validator_test.go]: Matches the merged #2065 comment (`upstreamDefinitions URLs must be host[:port] only; a path belongs in basePath`) and the comment in api_validator.go. Rewording only the new tests would make them inconsistent with the existing code, so keeping it as-is.

File: gateway/gateway-controller/pkg/config/validator_test.go [hunk 2]
```diff
@@ -620,6 +620,72 @@ func TestValidateUpstreamDefinitions_URLWithPathRejected(t *testing.T) {
...
+		},
+	}
+
+	errors := validator.validateUpstreamDefinitions(definitions)
+	require.Len(t, errors, 1)
+	assert.Equal(t, "spec.upstreamDefinitions[0].upstreams[0].url", errors[0].Field)
+	assert.Contains(t, errors[0].Message, "query string")
+}
+
+// upstreamDefinitions URLs must be host[:port] only; a bare "?" query marker is dropped, not honored.
```
[the remark this answers is not in the document]
Comment (mehara-rothila) [reply gateway/gateway-controller/pkg/config/validator_test.go]: Matches the merged #2065 comment (`upstreamDefinitions URLs must be host[:port] only; a path belongs in basePath`) and the comment in api_validator.go. Rewording only the new tests would make them inconsistent with the existing code, so keeping it as-is.

File: gateway/gateway-controller/pkg/config/validator_test.go [hunk 3]
```diff
@@ -620,6 +620,72 @@ func TestValidateUpstreamDefinitions_URLWithPathRejected(t *testing.T) {
...
+		},
+	}
+
+	errors := validator.validateUpstreamDefinitions(definitions)
+	require.Len(t, errors, 1)
+	assert.Equal(t, "spec.upstreamDefinitions[0].upstreams[0].url", errors[0].Field)
+	assert.Contains(t, errors[0].Message, "query string")
+}
+
+// upstreamDefinitions URLs must be host[:port] only; a fragment is dropped, not honored.
```
[the remark this answers is not in the document]
Comment (mehara-rothila) [reply gateway/gateway-controller/pkg/config/validator_test.go]: Matches the merged #2065 comment (`upstreamDefinitions URLs must be host[:port] only; a path belongs in basePath`) and the comment in api_validator.go. Rewording only the new tests would make them inconsistent with the existing code, so keeping it as-is.

File: gateway/gateway-controller/pkg/config/validator_test.go [hunk 4]
```diff
@@ -620,6 +620,72 @@ func TestValidateUpstreamDefinitions_URLWithPathRejected(t *testing.T) {
 	assert.Contains(t, errors[0].Message, "basePath")
 }
 
+// upstreamDefinitions URLs must be host[:port] only; a query string is dropped, not honored.
+func TestValidateUpstreamDefinitions_URLWithQueryRejected(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mehara-rothila) [reply gateway/gateway-controller/pkg/config/validator_test.go:624]: Keeping a separate function per case to match the rest of validator_test.go (including the #2065 sibling TestValidateUpstreamDefinitions_URLWithPathRejected). Table-driving only these three would diverge from the established pattern.

File: gateway/gateway-controller/pkg/config/validator_test.go [hunk 5]
```diff
@@ -620,6 +620,72 @@ func TestValidateUpstreamDefinitions_URLWithPathRejected(t *testing.T) {
...
+	}
+
+	errors := validator.validateUpstreamDefinitions(definitions)
+	require.Len(t, errors, 1)
+	assert.Equal(t, "spec.upstreamDefinitions[0].upstreams[0].url", errors[0].Field)
+	assert.Contains(t, errors[0].Message, "query string")
+}
+
+// upstreamDefinitions URLs must be host[:port] only; a bare "?" query marker is dropped, not honored.
+func TestValidateUpstreamDefinitions_URLWithBareQueryRejected(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mehara-rothila) [reply gateway/gateway-controller/pkg/config/validator_test.go:646]: Keeping a separate function per case to match the rest of validator_test.go (including the #2065 sibling TestValidateUpstreamDefinitions_URLWithPathRejected). Table-driving only these three would diverge from the established pattern.

File: gateway/gateway-controller/pkg/config/validator_test.go [hunk 6]
```diff
@@ -620,6 +620,72 @@ func TestValidateUpstreamDefinitions_URLWithPathRejected(t *testing.T) {
...
+	}
+
+	errors := validator.validateUpstreamDefinitions(definitions)
+	require.Len(t, errors, 1)
+	assert.Equal(t, "spec.upstreamDefinitions[0].upstreams[0].url", errors[0].Field)
+	assert.Contains(t, errors[0].Message, "query string")
+}
+
+// upstreamDefinitions URLs must be host[:port] only; a fragment is dropped, not honored.
+func TestValidateUpstreamDefinitions_URLWithFragmentRejected(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mehara-rothila) [reply gateway/gateway-controller/pkg/config/validator_test.go:668]: Keeping a separate function per case to match the rest of validator_test.go (including the #2065 sibling TestValidateUpstreamDefinitions_URLWithPathRejected). Table-driving only these three would diverge from the established pattern.

File: gateway/gateway-controller/pkg/config/api_validator.go [hunk 7]
```diff
@@ -341,6 +341,22 @@ func (v *APIValidator) validateUpstreamDefinitions(definitions *[]api.UpstreamDe
 						Message: "URL must not include a path; set the base path in upstreamDefinitions[].basePath instead",
 					})
 				}
+
+				// A query string or fragment is not part of the upstream cluster
+				// (host[:port] only), so it would be silently dropped — reject it.
+				if parsedURL.RawQuery != "" || parsedURL.ForceQuery {
+					errors = append(errors, ValidationError{
+						Field:   fmt.Sprintf("spec.upstreamDefinitions[%d].upstreams[%d].url", i, j),
+						Message: "URL must not include a query string",
```
[the remark this answers is not in the document]
Comment (mehara-rothila) [reply gateway/gateway-controller/pkg/config/api_validator.go]: Addressed in d5bc0b535 — the messages now explain why, e.g. "URL must not include a query string; it is not honored for upstreamDefinitions URLs".

**Comment by mehara-rothila** [comment]
@coderabbitai Can you do a final review and tell me if this PR is complete and ready to merge, or if anything is missing?

Context , this PR rejects a query string or fragment in `upstreamDefinitions[].upstreams[].url` (they get dropped when the cluster is built, so we fail fast at deploy with a 400 instead of silently ignoring them). Since your earlier reviews I:

- reworded the two validation messages to state the rule: `URL must not include a query string/fragment; only host[:port] is used`
- added an integration test in `api_deploy.feature` that deploys a query-string upstream URL and asserts the 400 plus the message
- the base was moved from `feature/operation-level-ep` to `main` (diff is still only the two validator files + the IT)

Please confirm:

1. Validation logic , does it fully cover query (`RawQuery` and the bare-`?` `ForceQuery`) and fragment, no edge cases missed?
2. Reworded messages , clear and consistent with the existing path-rejection message?
3. The new integration test , correct and enough, or should it also cover the fragment / bare-`?` case end to end?
4. Anything else blocking merge?

**Review by VirajSalaka — DISMISSED** [review]

File: gateway/it/features/api_deploy.feature [hunk 8]
```diff
@@ -128,3 +128,132 @@ Feature: API Deployment and Invocation
...
+        displayName: Invalid-Upstream-Bare-Query-API
+        version: v1.0
+        context: /invalid-upstream-bare-query/$version
+        vhosts:
+          main: invalid-upstream-bare-query.local
+        upstreamDefinitions:
+          - name: backend-default
+            basePath: /api-main
+            upstreams:
+              - url: http://sample-backend:9080?
```
Comment (renuka-fernando) [inline gateway/it/features/api_deploy.feature:182]: I think for these kinds of issues, we don't need gateway IT tests. Just unit tests would be enough.

**Review by renuka-fernando — APPROVED** [review]
